### ISSUE TITLE
DllImport issue for Linux

### ISSUE THREAD

**Issue by nonunknown** [mono/mono#20066]
Steps to Reproduce

1. Create a Hello world script using c++
2. Compile to linux lib `.so`
3. Import into Csharp:
![Screenshot from 2020-07-03 13-41-40](https://user-images.githubusercontent.com/58845030/86488895-e8d03980-bd38-11ea-9ed2-a300137b9665.png)

### Current Behavior

![Screenshot from 2020-07-03 14-22-51](https://user-images.githubusercontent.com/58845030/86488841-b6264100-bd38-11ea-8d66-16dd42a610e9.png)

### Expected Behavior

Console log: Hello World!

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:
Mono JIT compiler version 6.10.0.104 (tarball Fri Jun 26 19:38:44 UTC 2020)

## Stacktrace

![Screenshot from 2020-07-03 13-41-21](https://user-images.githubusercontent.com/58845030/86488945-0b625280-bd39-11ea-83a3-769ef9273104.png)

**Comment by HinTak** [comment]
Afaik, dllimport does not take absolute paths - you specify the file name only, and use the standard os mechanism (i.e. Setting LD_LIBRARY_PATH) for libraries.

**Comment by HinTak** [comment]
And you need to do extern c in your library too.

**Comment by nonunknown** [comment]
@HinTak I've tried all these possibilites:

external 'c'
dclspec()
DllImport ("code.so") // @"code.so"
exporting the folder containing the .so in the LD_LIBRARY_PATH 

All of those gave me same error: unknown assembly

**Comment by HinTak** [comment]
Try setting MONO_LOG_LEVEL=debug and see

**Comment by HinTak** [comment]
Hmm, I believe ".so" ( and ".dll") are auto-appended. Try removing it. And perhaps follow the platform convention - use `"code"` in your dllimport, and name your actual library `libcode.so`.

**Comment by nonunknown** [comment]
I'll try here, but about the lib prefix, I've tried this, mono automatic do this, so if I rename to libcode.so then it will read as "liblibcode.so"

**Comment by HinTak** [comment]
I mean you should have `[DllImport("code")]` when your library file is `libcode.so`.

**Comment by nonunknown** [comment]
ok lets suppose this will work, when I export the mono program for public, it will work without people needing to do the same thing?

**Comment by HinTak** [comment]
Yes, people need to do the same thing. Unless you use mkbundle. 

You don't seem to be familar with P/Invoke on Window?

**Comment by HinTak** [comment]
Here is a shipping example https://github.com/HinTak/Font-Validator/blob/0fd31052f7b9ecb26af2641057149c817314dc93/Compat/Compat.cs#L57

**Comment by nonunknown** [comment]
Never tried DLLImport on Windows, Trying first time with linux, cuz I'm porting a application to linux.

**Comment by HinTak** [comment]
You need you read about p/invoke https://www.mono-project.com/docs/advanced/pinvoke/

**Comment by nonunknown** [comment]
Can you provide for me a example cpp project that uses use extern C or dclspec to export the methods?

**Comment by HinTak** [comment]
You cannot pass class::method through p invoke - only plain functions.

**Comment by nonunknown** [comment]
well I'm trying to use Memory Read/Write with c# on linux, this is what brings me here, find a Cpp way to do write/read in linux and port to c#

**Comment by HinTak** [comment]
Let's be honest, you do not have a problem with P/Invoke on linux - you just don't know how to do it on windows either.

**Comment by HinTak** [comment]
Here is another pure c# example - http://www.psilocybegames.com/captainslog/2018/03/10/saving-and-loading-an-array-in-mono-c-in-unity-with-compression/

Did you even read any online docs above or by searching?

**Comment by nonunknown** [comment]
Man, There are many years I do not use windows, also If I was using it, this problem would never exists do you agree? this is linux related, Read/Write Memory Class works only with windows cuz the kernel32.dll, you cant import that on linux, so the only way I found, was to read cpp on C# mono. which requires a .so file in the DllImport.

**Comment by HinTak** [comment]
For the 3Rd time, stop talking about cpp - P/invoke only works for plain functions. Everything must be `extern "C"` . http://www.mono-project.com/Interop_with_Native_Libraries is the authoritative doc. You should just provide a snippet of you r kernel32 calling code and ask for the linux equivalent.

**Comment by HinTak** [comment]
If you are looking for `CopyMemory`, `MoveMemory`, `ZeroMemory` etc in kernel32, the equivalent in linux, don't bother. They are built-in 
to mono itself - you don't need to do anything. That's what internal means: https://gitlab.uwe.ac.uk/k9-hill/Portfolio/blob/8edb4b68a751891b9341c9d4a0d4cc6111654dcd/Unity-1stYear-UniProj-Stealth/Stealth_Data/Mono/etc/mono/config
